imap/caldav_alarm.c

```diff
@@ -1233,14 +1235,54 @@ EXPORTED int caldav_alarm_process(time_t runtime, time_t *intervalp)
 
     caldav_alarm_close(alarmdb);
 
-    int i;
-    for (i = 0; i < rock.list.count; i++) {
-        struct caldav_alarm_data *data = ptrarray_nth(&rock.list, i);
-        process_one_record(data, runtime);
-        caldav_alarm_fini(data);
-        free(data);
-    }
+    int next = 0;
+    char *prev_mboxname = NULL;
+    do {
+        int i = next, cnt = 0;
+
+        next = 0;  /* reset next pass start point */
+
+        for (; i < rock.list.count; i++) {
+            struct caldav_alarm_data *data = ptrarray_nth(&rock.list, i);
+
+            if (!data) continue;
+
+            /* XXX  we can do this comparison because the list is sorted */
+            if (!prev_mboxname ||
+                strncmp(data->mboxname, prev_mboxname, strlen(prev_mboxname))) {
+                /* different user/tree - reset counter */
+                cnt = 0;
+            }
+            else if (cnt >= MAX_CONSECUTIVE_ALARMS_PER_USER) {
```

Comment:
My spidey senses detect that this is mis-named, and should be `MAX_CONSECUTIVE_ALARMS_PER_MAILBOX`.

Reply:
I don't think so.  We only reset the counter when we hit a mailbox that is no longer a child or its predecessor.

But now that I look at this again, I think that logic is faulty.  If we have:
user.foo.A
user.foo.A.B
user.foo.B

This will reset the counter at user.foo.B, which is not what we want

Reply:
We should just check if the owner of the mailbox has changed, which is what I think I did originally before I tried to get too cute.

Reply:
Meh, I think counting by mailbox is reasonable - most people won't have tons of mailboxes each with tons of alarms.